InterruptedExceptions should never be ignored in the code.

kamaci opened the pull request.

leventov commented:
Thank you for this PR. Could we add some automatic checks? Does IntelliJ have some inspections about this?

At least `catch \([^)]*InterruptedException[^)]*\)\s*\{\s*throw` regexp could be added to Checkstyle to find it pretty effectively (especially if #6701 is also fixed.)

leventov commented on extensions-contrib/druid-rocketmq/src/main/java/org/apache/druid/firehose/rocketmq/RocketMQFirehoseFactory.java:
```diff
@@ -498,6 +502,7 @@ public void run()
       }
       catch (InterruptedException e) {
         LOGGER.error(e, "");
+        Thread.currentThread().interrupt();
```
I think it's always better to put this statement first, before logging. Because if an exception is thrown from logging the interruption status is shallowed.

kamaci replied:
I've updated my PR with a new commit to place such lines at first line at their corresponding code blocks.

leventov replied:
Let's assume that this PR is blocked by #6939.

jihoonson commented on core/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:
```diff
@@ -86,6 +86,7 @@ public int getPoolSize()
       return wrapObject(timeoutMs > 0 ? pollObject(timeoutMs) : pollObject());
     }
     catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
How about just propagate InterruptedException to callers? It makes more sense to me for all methods of BlockingPool.

kamaci replied:
Do you mean using `throw Throwables.propagate(e);`?

jihoonson replied:
No, I mean making `InterruptedException` as a checked exception for all methods of BlockingPool.

jihoonson replied:
BTW, I'm fine by making another PR for this change. This PR is still worth without it.

kamaci replied:
> BTW, I'm fine by making another PR for this change. This PR is still worth without it.

@leventov this PR is closed due to inactivity. However, it can be merged since https://github.com/apache/incubator-druid/pull/7121 has merged too?

leventov replied:
Yes. Please fix the conflicts.

jihoonson commented on core/src/main/java/org/apache/druid/java/util/http/client/pool/ResourcePool.java:
```diff
@@ -173,7 +173,7 @@ V get()
           }
           catch (InterruptedException e) {
-            Thread.interrupted();
+            Thread.currentThread().interrupt();
```
Nice catch!

kamaci commented:
@leventov I've resolved conflicts. We've accepted that this issue was blocked by #7121 and it has been merged. Is there anything left to improve this PR?

leventov commented:
@kamaci please add a Checkstyle check a-la https://github.com/apache/incubator-druid/pull/6943#issuecomment-458534551. It may be also possible to catch such patterns more effectively with [Structural Search](https://github.com/apache/incubator-druid/pull/7275), but I didn't experiment with that.

kamaci commented:
> @kamaci please add a Checkstyle check a-la [#6943 (comment)](https://github.com/apache/incubator-druid/pull/6943#issuecomment-458534551). It may be also possible to catch such patterns more effectively with [Structural Search](https://github.com/apache/incubator-druid/pull/7275), but I didn't experiment with that.

@leventov I've added the checkstyle rule as a warning since it may not be a violation not to obey it.

leventov commented:
@kamaci in which case this is not a violation? Could you please point to an example?

kamaci commented:
> @kamaci in which case this is not a violation? Could you please point to an example?

@leventov  I've removed that style from being warning. On the other hand, I've edited checkstyle regex for not to fail on `QueryInterruptedException`. I've also fixed test classes about not to ignore `InterruptedException`.

leventov commented on core/src/main/java/org/apache/druid/java/util/http/client/pool/ResourcePool.java:
```diff
@@ -186,7 +186,7 @@ V get()
             this.wait();
           }
           catch (InterruptedException e) {
-            Thread.interrupted();
```
Perhaps we should prohibit this method on the forbidden-apis level?

kamaci replied:
I agree! One can call `Thread.interrupted()` within an `if` check or assign it to a variable. Should we prohibit to assigning a variable too or not?

    <module name="Regexp">
      <property name="format" value='Thread.interrupted();'/>
      <property name="illegalPattern" value="true"/>
      <property name="message" value="Thread.interrupted(); shouldn't be used individually"/>
    </module>

leventov replied:
It should be prohibited via forbidden-apis, not Checkstyle.

> Should we prohibit to assigning a variable too or not?

Do you mean prohibiting `Thread#interrupt()`? It's called in a lot of places.

kamaci replied:
I mean we do not forbid to use Thread.interrupted() i.e. these are valid usages:

      boolean needsToShutdown = Thread.interrupted() || shuttingDown;
      ...
      if (Thread.interrupted()) {
        throw new InterruptedException();
      }

It can be assigned to a variable or can be used within an `if` condition. So, we should only prohibit to use it without assigning to a variable or not using inside an `if` condition.

leventov replied:
Ok. Maybe you can achieve this with IntelliJ's structural search inspections: https://github.com/apache/incubator-druid/blob/78fd5aff21f692f583955a281e841b1e8fd6925f/ci/README_TeamCity.md#creating-a-custom-inspection-from-a-structural-search-pattern

leventov replied:
@kamaci could you please prohibit using Checkstyle like
```
^\s+Thread\.interrupted\(\)\;
```

kamaci commented:
@leventov @jihoonson I've resolved the conflicts and all checks have passed.

leventov commented on indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:
```diff
@@ -761,6 +761,7 @@ public void tryInit()
               catch (InterruptedException e) {
                 stateManager.recordThrowableEvent(e);
                 log.info("SeekableStreamSupervisor[%s] interrupted, exiting", dataSource);
+                Thread.currentThread().interrupt();
```
By convention which you use in this PR, `Thread.currentThread().interrupt();` (optionally wrapped in `if (e instanceof InterruptedException)`) is the first statement in the catch block; please move it here above, too. Also in other places in this file `SeekableStreamSupervisor.java`.

leventov commented on sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java:
```diff
@@ -305,6 +305,7 @@ public void run()
             }
             catch (InterruptedException e) {
               // Just exit.
+              Thread.currentThread().interrupt();
```
Is it aligned with the semantics of "Just exit"? As a minimum, the comment should go after, not before the `Thread.currentThread().interrupt();` statement.

leventov commented:
@kamaci could you please complete this?

kamaci commented:
> @kamaci could you please complete this?

Sure, I will.
